### PR TITLE
Decoder Only Generation Support

### PR DESCRIPTION
**Pull request by jonathangomesselman**
## Description

In this PR we focus on adding generation support for Decoder Only models. In doing so, we also re-structure parts of the generation workflow. Below is an executive summary of the changes added!

The primary differences between generation for EncoderDecoder and DecoderOnly models are:
1. The df column / sample data used for generation
2. The generation processing for each sample (i.e. the logic previously in the `utils.seq2seq.generation.generate_sample_output` function).

Thankfully, similar to the processing of model logprob outputs, after sample level generation / processing, logic can be shared around token alignment and the overall saving of information to the dataframe - since in the end we expect the same information for generated data.

**The main structural change is to address (2) from above.**

To do this we leverage the `BaseSeq2SeqDataFormatter` class and define a new function `generate_sample` that is is over-written by `EncoderDecoderDataFormatter` and `DecoderOnlyDataFormatter`. Since generation occurs within `Seq2SeqDataLogger`, when we call `generation.add_generated_output_to_df` (from `Seq2SeqDataLogger.add_generated_output_to_df`), we now pass along the correct `formatter`. Then within `generation.generate_on_batch`, in place of calling `generate_sample_output` we call the respective `formatter.generate_sample`. 

The above description gives a high level intuition for the big changes in code structure. Additionally, there are some new parameters needed for generation with DecoderOnly models, such as the `sample_id` we are generating on (necessary to help isolate just the "input prompt" prompt to the model by removing the response tokens). 

**Addressing Difference (1)**
This is a fairly simple but important difference. 
- EncoderDecoder models can directly use the `Input` text column for generation.
- DecoderOnly models need the `formatted_prompt` text column to account for additional prompt formatting applied to the `Input` / `Target` pair.

**Additional Things to Note**
When first testing, I noticed that I was having a strange off by 1 error in the alignment. Essentially, the top-5 tokens for each "word" were actually the top-5 tokens for the next word. 

The problem ended up being that the tokenizer I was using added a `<bos>` token by default. However, for DecoderOnly models, although I am aligning just the generated tokens / string, I do so "as if" it were directly pre-fixed by the prompt. Therefore, I do not want any `<bos>` tokens added because I do not have logits for these, and thus the token alignments were shifted by one!

If the above does not immediately make much sense that is okay. The main takeaway is that within `utils.seq2seq.generation.generate_on_batch`, I explicitly decode and tokenize without adding *or* removing special tokens - i.e. I use exactly the generated ids - to avoid alignment issues. Additionally, this issue pointed me to potential challenges if tokenizers add <bos> tokens that we should test for when adding test completeness.

## Demonstration

You should now be able to generate using the `Seq2Seq_Decoder_Only_Dq_Fake_Data` notebook.

Here is an example [run](https://console.dev.rungalileo.io/insights/8de23502-d6cc-44a5-8452-0999ff6d6c92/cf1b9db4-fa8f-436f-8f2c-4b4af55406c5?dataframeColumns=bleu%3B%26data_error_potential%3B%26generated_output%3B%26generated_token_logprobs%3B%26generated_uncertainty%3B%26input%3B%26perplexity%3B%26rouge%3B%26target&groupedBy=galileo_input_length&taskType=8).

### CODE REVIEW

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py [hunk 1]
```diff
@@ -318,10 +323,29 @@ def add_generated_output_to_df(
...
+            generation_column=generation_column,
+            formatter=self.formatter,
+            tokenizer=tokenizer,
+            model=model,
+            max_input_tokens=max_input_tokens,
+            generation_config=generation_config,
+            split_key=split,
         )
+        # The formatted_prompts column is no longer needed
+        df = df.drop([S2SIC.formatted_prompts])
```
[the remark this answers is not in the document]
Comment (bogdan-galileo) [reply dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:347]: Fixing it 
Screenshot 2023-12-06 at 4 11 44 PM

**Comment by jonathangomesselman** [comment]
> It takes an unreasonable amount of time (almost 10 minutes !!!) on Screenshot 2023-12-06 at 4 32 37 PM
> 
> Tested end to end and it does generate
> 
> * enc_dec run https://console.dev.rungalileo.io/insights/e24eb571-8eaf-45a1-beb0-86c6488f4f1a/314f813f-8548-4b34-ab67-0e5e6b402c5c?split=validation&taskType=8
> * dec_only run https://console.dev.rungalileo.io/insights/e24eb571-8eaf-45a1-beb0-86c6488f4f1a/a30025a0-2526-42d8-a903-dabbe2f9b66f?split=training&taskType=8

@bogdan-galileo - Are you using just the sample notebook I created?

It certainly seems odd. I ran many local tests and for 10 samples with the facebook `opt-170m` model, it takes:
- max_tokens generated = 15 --> `55s` total
- max_tokens generated = 100 --> `42s` for generation and `111s` total

See this [run](https://console.dev.rungalileo.io/insights/8de23502-d6cc-44a5-8452-0999ff6d6c92/e87441ce-684d-426e-a042-accc7c5b505d?dataframeColumns=data_error_potential%3B%26generated_token_logprobs%3B%26input%3B%26perplexity%3B%26target&groupedBy=galileo_input_length&taskType=8) for `max_tokens generated = 100`.

**Comment by bogdan-galileo** [comment]
> > It takes an unreasonable amount of time (almost 10 minutes !!!) on Screenshot 2023-12-06 at 4 32 37 PM
> > Tested end to end and it does generate
> > 
> > * enc_dec run https://console.dev.rungalileo.io/insights/e24eb571-8eaf-45a1-beb0-86c6488f4f1a/314f813f-8548-4b34-ab67-0e5e6b402c5c?split=validation&taskType=8
> > * dec_only run https://console.dev.rungalileo.io/insights/e24eb571-8eaf-45a1-beb0-86c6488f4f1a/a30025a0-2526-42d8-a903-dabbe2f9b66f?split=training&taskType=8
> 
> @bogdan-galileo - Are you using just the sample notebook I created?
> 
> It certainly seems odd. I ran many local tests and for 10 samples with the facebook `opt-170m` model, it takes:
> 
> * max_tokens generated = 15 --> `55s` total
> * max_tokens generated = 100 --> `42s` for generation and `111s` total
> 
> See this [run](https://console.dev.rungalileo.io/insights/8de23502-d6cc-44a5-8452-0999ff6d6c92/e87441ce-684d-426e-a042-accc7c5b505d?dataframeColumns=data_error_potential%3B%26generated_token_logprobs%3B%26input%3B%26perplexity%3B%26target&groupedBy=galileo_input_length&taskType=8) for `max_tokens generated = 100`.

I used the notebook in `rungalileo/examples` for encoder-decoder and the one in the dq repo for decoder-only. But it seems to take a long time after dq work is done, which is why I'm so confused. Plat has been updated how runners job are queued and all, maybe it's related ?

File: dataquality/schemas/seq2seq.py [hunk 2]
```diff
@@ -38,6 +38,10 @@ class Seq2SeqInputCols(str, Enum):
     system_prompts = "system_prompts"
 
 
+class Seq2SeqInputTempCols(str, Enum):
```
Comment (bogdan-galileo) [inline dataquality/schemas/seq2seq.py:41]: @jonathangomesselman @elboy3 I created this class because this is a temporary column that will not be saved. If we put it in `Seq2SeqInputCols` we have the attribute test that fails. Let me know what you think

Comment (jonathangomesselman) [reply dataquality/schemas/seq2seq.py:41]: Definitely an @elboy3 question!

File: dataquality/loggers/data_logger/seq2seq/formatters.py [hunk 3]
```diff
@@ -71,6 +81,73 @@ def format_text(
...
+        to extract the token logprobs. We effectively ask the model
+        to evaluate its own generation - which is identical to generation
+        because of causal language modeling.
+
+        Parameters:
+        -----------
+        input_str: str
+            Input string context used to seed the generation
+        tokenizer: PreTrainedTokenizerFast
+        max_input_tokens: the max number of tokens to use for tokenization
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/formatters.py:115]:
```suggestion
        max_input_tokens: int
            the max number of tokens to use for tokenization
```

File: dataquality/loggers/data_logger/seq2seq/formatters.py [hunk 4]
```diff
@@ -71,6 +81,73 @@ def format_text(
...
+    @abstractmethod
+    def generate_sample(
+        self,
+        input_str: str,
+        tokenizer: PreTrainedTokenizerFast,
+        model: PreTrainedModel,
+        max_input_tokens: int,
+        generation_config: GenerationConfig,
+        input_id: Optional[int] = None,
+        split_key: Optional[str] = None,
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/formatters.py:94]: these are missing from docstring

File: dataquality/loggers/data_logger/seq2seq/formatters.py [hunk 5]
```diff
@@ -71,6 +81,73 @@ def format_text(
...
+        generation_config: GenerationConfig
+            Users generation config specifying the parameters for generation
+
+        Return:
+        -------
+        model_generation: ModelGeneration
+            - generated_ids: np.ndarray of shape - [seq_len]
+            - generated_token_logprobs: np.ndarray of shape - [seq_len]
+            - generated_top_logprobs: List[List[Tuple[str, float]]]
+        """
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/formatters.py:126]: nice!

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py [hunk 6]
```diff
@@ -88,14 +90,11 @@ def __init__(self, meta: Optional[MetasType] = None) -> None:
         self.ids: List[int] = []
         self.texts: List[str] = []
         self.labels: List[str] = []
-        # Only requied for Decoder-Only models
+        # Only required for Decoder-Only models
         self.formatted_prompts: List[str] = []
         # Formatter distinguishes behavior between EncoderDecoder and DecoderOnly
-        from dataquality.loggers.data_logger.seq2seq.formatters import (
-            BaseSeq2SeqDataFormatter,
-        )
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:96]: very nice, love removing import  from here

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py [hunk 7]
```diff
@@ -290,22 +298,21 @@ def convert_large_string(self, df: DataFrame) -> DataFrame:
...
-        cls, df: DataFrame, split: str
+        self, df: DataFrame, split: str
     ) -> Optional[DataFrame]:
         """Adds the generated output to the dataframe
         Adds the generated output to the dataframe, and also adds the
         `token_label_positions` column
         """
-        logger_config = cls.logger_config
-        if split not in logger_config.generation_splits:
+        # logger_config = cls.logger_config
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:308]: 🧹

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py [hunk 8]
```diff
@@ -318,10 +325,27 @@ def add_generated_output_to_df(
             )
         assert isinstance(max_input_tokens, int)
 
+        assert generation_config is not None
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:328]: this could be none if you watch without the config but pass in generation splits as not none? maybe we can do that validation in watch (if we don't already)

File: dataquality/schemas/seq2seq.py (hunk 2 above)
Comment (elboy3) [reply dataquality/schemas/seq2seq.py:41]: Works for me. Nit:  I would just call it `Seq2SeqTemporaryCols` and remove input and we can use it for input or output dfs

File: dataquality/utils/seq2seq/decoder_only.py [hunk 9]
```diff
@@ -26,42 +26,55 @@ def extract_tokenized_responses(
     If a sample does not contain the response_template we represent the
     tokenized_response for that sample as [] - i.e. the <Empty String>.
     """
+    # Reverse search over matches of the first token in the response template
+    matched_indices = np.where(
+        np.array(tokenized_formatted_prompt) == response_template[0]
+    )[0]
+    response_token_ids_start_idx = None
+    for i in range(len(matched_indices)):
+        match_idx = matched_indices[-(i + 1)]
```
Comment (elboy3) [inline dataquality/utils/seq2seq/decoder_only.py:35]: could we do some kidn of `for idx in reverse(matched_indices)` ?

File: dataquality/loggers/data_logger/seq2seq/formatters.py [hunk 10]
```diff
@@ -338,6 +457,63 @@ def format_text(
...
+        split_key: Optional[str] = None,
+    ) -> ModelGeneration:
+        """Generate response for a single sample - Decoder Only"""
+
+        # Retrieve the length of the response tokens
+        assert split_key, "Decoder-Only models require a valid `split_key`"
+        assert input_id is not None, (
+            "Decoder-Only models require the `sample_id` "
+            "that is being used for generation"
+        )
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/formatters.py:478]: Not asking for a change just noting - i feel like these assertion errors should guide the user on what they should fix. If these assertions fail, what went wrong?

File: dataquality/loggers/data_logger/seq2seq/formatters.py [hunk 11]
```diff
@@ -338,6 +457,63 @@ def format_text(
...
+        # Shape - [1, seq_len]
+        # We make sure to tokenize with the same max_length as they did during training
+        formatted_prompt_ids = tokenizer(
+            input_str,
+            truncation=True,
+            max_length=max_input_tokens,
+            verbose=False,
+            return_tensors="pt",
+        )["input_ids"]
+        # TODO Watch for empty targets!
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/formatters.py:491]: what's the TODO action here?

File: dataquality/loggers/data_logger/seq2seq/formatters.py [hunk 12]
```diff
@@ -338,6 +457,63 @@ def format_text(
...
+        just_prompt_ids = formatted_prompt_ids[:, :-num_response_labels].to(
```
Comment (elboy3) [inline dataquality/loggers/data_logger/seq2seq/formatters.py:492]: just_prompt_ids feels like an odd var name.

Comment (elboy3) [reply dataquality/loggers/data_logger/seq2seq/formatters.py:492]: `prompt_ids_without_response` ? maybe not better 🤷‍♀️

**Review by elboy3 — APPROVED** [review]
Looking good. I'm approving but I think there's some cleanup (on prints, assertion statements, and comments) that could be done. 

Stellar!

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py (hunk 6 above)
Comment (jonathangomesselman) [reply dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:96]: Yeah I think it was necessary haha

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py [hunk 13]
```diff
@@ -280,7 +283,12 @@ def convert_large_string(self, df: DataFrame) -> DataFrame:
         See BaseDataLogger.convert_large_string for more details
         """
         df_copy = df.copy()
-        for text_col in [S2SIC.input.value, S2SIC.target.value]:
+        # TODO Include formatted prompt?
```
[the remark this answers is not in the document]
Comment (jonathangomesselman) [reply dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:286]: Wait what is this function doing? When did you add this?

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py (hunk 8 above)
Comment (jonathangomesselman) [reply dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:328]: Oh I see! But wait could the formatter ever be None?

File: dataquality/utils/seq2seq/decoder_only.py (hunk 9 above)
Comment (jonathangomesselman) [reply dataquality/utils/seq2seq/decoder_only.py:35]: Yeah 100% - I didn't really think about this!

File: dataquality/loggers/data_logger/seq2seq/formatters.py (hunk 10 above)
Comment (jonathangomesselman) [reply dataquality/loggers/data_logger/seq2seq/formatters.py:479]: Since this is not really something the user would mess up should we just have no error message? This is mostly for typing

File: dataquality/loggers/data_logger/seq2seq/formatters.py (hunk 11 above)
Comment (jonathangomesselman) [reply dataquality/loggers/data_logger/seq2seq/formatters.py:492]: Oh I think because [:, :-0] would be bad, since it would give none of the ids - but we actually want the whole thing! I will fix this.

Comment (jonathangomesselman) [reply dataquality/loggers/data_logger/seq2seq/formatters.py:492]: I added a fix that should handle this case seamlessly

**Comment by jonathangomesselman** [comment]
Thanks for all the great comments @elboy3 @bogdan-galileo - I noticed on or two bugs that must have snuck in with other cleanup changes. So I also have fixed those. Let's get this in next week!
